bitnami/harbor/templates/clair/clair-dpl.yaml

```
@@ -18,7 +18,13 @@ spec:
     matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
       app.kubernetes.io/component: clair
   {{- if .Values.clair.updateStrategy }}
-  strategy: {{- toYaml .Values.clair.updateStrategy | nindent 4 }}
+  strategy:
+    type: {{ .Values.clair.updateStrategy.type }}
+    {{- if (eq "Recreate" .Values.clair.updateStrategy.type) }}
+    rollingUpdate: null
+    {{- else }}
+    rollingUpdate: {{ toYaml .Values.clair.updateStrategy.rollingUpdate | nindent 6 }}
+    {{- end }}
```

To be compatible with #9533 , I think all these block could be like this:
```suggestion
  strategy:
    type: {{ .Values.clair.updateStrategy.type }}
    {{- if (eq "Recreate" .Values.clair.updateStrategy.type) }}
    rollingUpdate: {{ toYaml .Values.clair.updateStrategy.rollingUpdate | nindent 6 }}
    {{- end }}
```
What do you think ?

---

I think it was already compatible with that. When the type is `Recreate` what I added initially will set the `rollingUpdate` to `null`. If we apply your suggestion it won't be compatible

---

There's literally no reason for this if-statement at all if we just set `rollingUpdate` to `null` in `values.yaml`.

---

We must support the `rollingUpdate` configuration.
`null` is not the default value for any type. Setting it to `null` in the `values.yaml` should not be an alternative, nor deleting the parameter.

---

At the very least, I think the conditional logic should be rewritten to better reflect the Kubernetes API documentation:

> Present only if DeploymentStrategyType = RollingUpdate.

In my opinion, that should be interpreted/expressed in our template logic as

> if the type is `RollingUpdate`, then specify `rollingUpdate`

rather than

> if the type is `Recreate`, then do not specify `rollingUpdate`; otherwise, specify `rollingUpdate`

That should make it more robust with respect to any future changes to the Kubernetes API, and it's closer to what we *mean*™.

So I would flip the logic around and remove the `else` branch for clarity and simplicity:

```suggestion
  strategy:
    type: {{ .Values.clair.updateStrategy.type }}
    {{- if (eq "RollingUpdate" .Values.clair.updateStrategy.type) }}
    rollingUpdate: {{ toYaml .Values.clair.updateStrategy.rollingUpdate | nindent 6 }}
    {{- end }}
```

---

> `null` is not the default value for any type.

I think a clarification on this arguably hard-to-falsify claim might be interesting and helpful to understanding the issue at hand. What type space are we talking about, for starters?

* YAML types? Do they even have default values? Please also note that [`null` is a type according to the specification](https://yaml.org/spec/1.2.2/#24-tags), as well as the only value of that type.
* Go types? `null` is not a Go literal, but `nil` is the default (zero) value of all pointer types.
* Some JSON-schema type space? What is the connection between it and our template files?
* Some more abstract type space? How are its default values defined?